x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/SearchableSnapshotDirectory.java

```diff
@@ -480,77 +477,66 @@ private void prewarmCache(ActionListener<Void> listener) {
-                final IndexInput input = openInput(file.physicalName(), CachedBlobContainerIndexInput.CACHE_WARMING_CONTEXT);
-                assert input instanceof CachedBlobContainerIndexInput : "expected cached index input but got " + input.getClass();
-
-                final int numberOfParts = file.numberOfParts();
-                final StepListener<Void> fileCompletionListener = new StepListener<>();
-                fileCompletionListener.addListener(completionListener);
-                fileCompletionListener.whenComplete(ignored -> {
-                    logger.debug("{} file [{}] prewarmed", shardId, file.physicalName());
-                    input.close();
-                }, e -> {
-                    logger.warn(() -> format("%s prewarming failed for file [%s]", shardId, file.physicalName()), e);
-                    IOUtils.closeWhileHandlingException(input);
-                });
-
-                final CountDownActionListener partsListener = new CountDownActionListener(numberOfParts, fileCompletionListener);
-                submitted = true;
-                for (int p = 0; p < numberOfParts; p++) {
-                    final int part = p;
-                    queue.add(Tuple.tuple(partsListener, () -> {
-                        ensureOpen();
-
-                        logger.trace("{} warming cache for [{}] part [{}/{}]", shardId, file.physicalName(), part + 1, numberOfParts);
-                        final long startTimeInNanos = statsCurrentTimeNanosSupplier.getAsLong();
-                        final long persistentCacheLength = ((CachedBlobContainerIndexInput) input).prefetchPart(part).v1();
-                        if (persistentCacheLength == file.length()) {
-                            recoveryState.markIndexFileAsReused(file.physicalName());
-                        } else {
-                            recoveryState.getIndex().addRecoveredFromSnapshotBytesToFile(file.physicalName(), file.partBytes(part));
+                recoveryState.getIndex().addFileDetail(file.physicalName(), file.length(), false);
+                try {
+                    final IndexInput input = openInput(file.physicalName(), CachedBlobContainerIndexInput.CACHE_WARMING_CONTEXT);
+                    assert input instanceof CachedBlobContainerIndexInput : "expected cached index input but got " + input.getClass();
+
+                    try (
+                        var fileListener = new RefCountingListener(
+                            ActionListener.runBefore(completionListener.acquire(), () -> IOUtils.closeWhileHandlingException(input))
+                        )
+                    ) {
+                        for (int p = 0; p < file.numberOfParts(); p++) {
+                            final int part = p;
+                            queue.add(
+                                ActionRunnable.supply(
+                                    ActionListener.runAfter(fileListener.acquire(), () -> prewarmNext(executor, queue)),
```

Comment:
nit: I think we could use a `ThrottledTaskRunner` rather than explicitly having to call `prewarmNext` here.

Reply:
Good suggestion, I pushed 7afcb5b8f20f4144a83dfc1d7eeeb1c974baa1c2, let me know what you think